airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/report_streams.py

```diff
@@ -263,51 +268,65 @@ def _send_http_request(self, url: str, profile_id: int, json: dict = None):
+        if saved_date:
+            saved_date = pendulum.from_format(saved_date, self.REPORT_DATE_FORMAT).date()
+            today = pendulum.today(tz=profile.timezone).date()
+            if today.subtract(days=self.LOOK_BACK_WINDOW) < saved_date <= today:
+                return today.subtract(days=self.LOOK_BACK_WINDOW)
+            return saved_date
+
+    def get_start_report_date(self, profile: Profile, stream_state: Mapping[str, Any]) -> Date:
+        start_date = self.get_start_date(profile, stream_state)
+        saved_date = self.get_saved_date(profile, stream_state)
+        if saved_date:
+            return max(start_date, saved_date)
+        return start_date
 
     def stream_slices(
         self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
     ) -> Iterable[Optional[Mapping[str, Any]]]:
 
-        if sync_mode == SyncMode.full_refresh:
-            # For full refresh stream use date from config start_date field.
-            start_date = self._start_date
-        else:
-            # incremental stream
-            stream_state = stream_state or {}
-            start_date = stream_state.get(self.cursor_field)
-            if start_date:
-                start_date = pendulum.from_format(start_date, ReportStream.REPORT_DATE_FORMAT, tz="UTC")
-                start_date += timedelta(days=-ReportStream.LOOK_BACK_WINDOW)
-
-            else:
-                start_date = self._start_date
+        stream_state = stream_state or {}
 
-        return [{self.cursor_field: date} for date in ReportStream.get_report_date_ranges(start_date)] or [None]
+        slices = []
+        for profile in self._profiles:
+            today = pendulum.today(tz=profile.timezone).date()
+            start_report_date = self.get_start_report_date(profile, stream_state)
+            for report_date in self.get_report_date_ranges(start_report_date, today):
+                slices.append({"profile": profile, self.cursor_field: report_date})
+        if not slices:
+            return [None]
+        return slices
 
     def get_updated_state(self, current_stream_state: Dict[str, Any], latest_data: Mapping[str, Any]) -> Mapping[str, Any]:
-        return {"reportDate": latest_data["reportDate"]}
+        profileId = str(latest_data["profileId"])
+        updated_state = latest_data[self.cursor_field]
+        stream_state_value = current_stream_state.get(profileId, {}).get(self.cursor_field)
+        if stream_state_value:
+            updated_state = max(updated_state, stream_state_value)
+        current_stream_state.setdefault(profileId, {})[self.cursor_field] = updated_state
+
+        # store "start_date" in state
+        if not self._start_date and not current_stream_state.get(self.STATE_START_DATE_KEY):
+            current_stream_state[self.STATE_START_DATE_KEY] = pendulum.now("UTC").replace(microsecond=0).to_iso8601_string()
```

Comment:
is there a reason not to set this to `get_start_date(profileId, current_stream_state)` If I understand correctly, that method should return the stat's start date

Reply:
I have think today more and decided to re-implement `get_start_date` 
now `get_start_date` pretty simple method 
but more logic moved to `get_updated_state`